src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
-            out.writeInt(0);
+    private int calculateInstanceDumpRecordSize(Instance instance) {
+        Klass klass = instance.getKlass();
+        if (klass.getClassLoaderData() == null) {
+            // Ignoring this object since the corresponding Klass is not loaded.
+            // Might be a dormant archive object.
+            return 0;
+        }
+
+        ClassData cd = (ClassData) classDataCache.get(klass);
+        if (Assert.ASSERTS_ENABLED) {
+            Assert.that(cd != null, "can not get class data for " + klass.getName().asString() + klass.getAddress());
+        }
+        List<Field> fields = cd.fields;
+        return (int) BYTE_SIZE + (int)OBJ_ID_SIZE * 2 + (int)INT_SIZE * 2 + getSizeForFields(fields);
+    }
+
+    private int calculateClassDumpRecordSize(Klass k) throws IOException {
+        // tag + javaMirror + DUMMY_STACK_TRACE_ID + super
+        int size = (int)BYTE_SIZE + (int)INT_SIZE + (int)OBJ_ID_SIZE * 2;
+        if (k instanceof InstanceKlass) {
+            InstanceKlass ik = (InstanceKlass) k;
+            List<Field> fields = getInstanceFields(ik);
+            List<Field> declaredFields = ik.getImmediateFields();
+            List<Field> staticFields = new ArrayList<>();
+            List<Field> instanceFields = new ArrayList<>();
+            Iterator<Field> itr = null;
+            // loader + signer + protectionDomain + 2 reserved + fieldSize + cpool entris number
+            size += OBJ_ID_SIZE * 5 + INT_SIZE + SHORT_SIZE;
+            for (itr = declaredFields.iterator(); itr.hasNext();) {
+                Field field = itr.next();
+                if (field.isStatic()) {
+                    staticFields.add(field);
+                } else {
+                    instanceFields.add(field);
+                }
+            }
+            // size of static field descriptors
+            size += calculateFieldDescriptorsDumpRecordSize(staticFields, ik);
+            // size of instance field descriptors
+            size += calculateFieldDescriptorsDumpRecordSize(instanceFields, null);
+        } else {
+            size += OBJ_ID_SIZE * 5  + INT_SIZE + SHORT_SIZE * 3;
         }
+        return size;
+    }
+
+    private int calculateFieldDescriptorsDumpRecordSize(List<Field> fields, InstanceKlass ik)
+            throws IOException {
+        int size = 0;
+        size += SHORT_SIZE;
+        for (Iterator<Field> itr = fields.iterator(); itr.hasNext();) {
+            Field field = itr.next();
+            char typeCode = (char) field.getSignature().getByteAt(0);
+            size += OBJ_ID_SIZE + BYTE_SIZE;
+            // ik == null for instance fields
+            if (ik != null) {
+                // static field
+                size += calculateFieldDumpRecordSize(field, ik.getJavaMirror());
+            }
+        }
+        return size;
+    }
+
+    private int calculateFieldDumpRecordSize(Field field, Oop oop) throws IOException {
```

Comment:
This method is very similar to getSizeForFields
I think it makes sense to introduce method getSizeForField(Field) and use it here and in getSizeForFields
And I'd simplified the code to something like
```suggestion
    private int getSizeForField(Field field)  {
        char typeCode = (char)field.getSignature().getByteAt(0);
        switch (typeCode) {
            case JVM_SIGNATURE_BOOLEAN:
            case JVM_SIGNATURE_BYTE:
                return 1;
            case JVM_SIGNATURE_CHAR:
            case JVM_SIGNATURE_SHORT:
                return 2;
            case JVM_SIGNATURE_INT:
            case JVM_SIGNATURE_FLOAT:
                return 4;
            case JVM_SIGNATURE_CLASS:
            case JVM_SIGNATURE_ARRAY:
                return OBJ_ID_SIZE;
            case JVM_SIGNATURE_LONG:
            case JVM_SIGNATURE_DOUBLE:
                return 8;
            default:
                throw new RuntimeException("should not reach here");
        }
    }

```